### PR TITLE
Speed of CommandsCache increased in case when aliases have multiple updates (i.e. conda init).

### LINKED ISSUE #3912: Slower start: read files from PATH dirs many times
Hi! xonsh is shell but python but shell but python but the best!

While I worked on #3911 I found that reading of the files in the PATH directories was done many times in case of using default conda init:

```bash
pip install -U git+https://github.com/anki-code/xonsh.git@wsl_speed

# Run without rc - 1 warning
XONSH_WARNING_PATH_FILES=10 xonsh --no-rc
#Warning! Found 5,785 executable files in the PATH directories!
```
```bash
cat ~/.xonshrc_conda
# >>> conda initialize >>>
# !! Contents within this block are managed by 'conda init' !!
import sys as _sys
from types import ModuleType as _ModuleType
_mod = _ModuleType("xontrib.conda",
                   "Autogenerated from $(/opt/miniconda/bin/conda shell.xonsh hook)")
__xonsh__.execer.exec($("/opt/miniconda/bin/conda" "shell.xonsh" "hook"),
                      glbs=_mod.__dict__,
                      filename="$(/opt/miniconda/bin/conda shell.xonsh hook)")
_sys.modules["xontrib.conda"] = _mod
del _sys, _mod, _ModuleType
# <<< conda initialize <<<
```
```bash
# Run with conda rc - 3 warnings
XONSH_WARNING_PATH_FILES=10 xonsh --rc ~/.xonshrc_conda 
#Warning! Found 5,785 executable files in the PATH directories!
#Warning! Found 5,785 executable files in the PATH directories!
#Warning! Found 5,785 executable files in the PATH directories!
```

**As result instead of scan 5795 - 2298 - 420 = 3077 files once it scans 5795 files 3 times = 17,385 in sum.**
It looks like this is the cause why xonsh have 2 sec start time on my linux!

Please take a look at `commands_cache.py` and conda init. It looks like a speed issue.

Also `--timings` not shows the time of creating prompt on the first run that uses commands_cache so this is why no one can detect the problem.

## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

### PR DESCRIPTION
**Pull request by anki-code**
Prerequisites:
* mostly #3912
* a bit of #3915

Before this PR (values after 10 runs):
```
$ xonsh --timings --rc ~/.xonshrc_conda 
*
*
*
|------------------|-----------|-----------|
|Event name        | Time (s)  | Delta (s) |
|------------------|-----------|-----------|
|start             |   0.000   |   0.000   |
|pre_execer_init   |   0.015   |   0.015   |
|post_execer_init  |   0.057   |   0.042   |
|on_pre_rc         |   0.057   |   0.000   |
|on_post_rc        |   0.253   |   0.196   |
|on_ptk_create     |   0.334   |   0.081   |
|on_post_init      |   0.334   |   0.000   |
|on_pre_cmdloop    |   0.334   |   0.000   |
|on_pre_prompt     |   0.738   |   0.403   |
|------------------|-----------|-----------|
```

After this PR (values after 10 runs):
```
$ xonsh --timings --rc ~/.xonshrc_conda 
*
|------------------|-----------|-----------|
|Event name        | Time (s)  | Delta (s) |
|------------------|-----------|-----------|
|start             |   0.000   |   0.000   |
|pre_execer_init   |   0.014   |   0.014   |
|post_execer_init  |   0.057   |   0.043   |
|on_pre_rc         |   0.057   |   0.000   |
|on_post_rc        |   0.223   |   0.166   |
|on_ptk_create     |   0.306   |   0.083   |
|on_post_init      |   0.306   |   0.000   |
|on_pre_cmdloop    |   0.306   |   0.000   |
|on_pre_prompt     |   0.389   |   0.082   |
|------------------|-----------|-----------|
```

As result: 0.389 vs 0.738 = 50% speed up on Linux.

Note! To see this speed up you should use #3916.

Closes #3912

### CODE REVIEW

**Comment by anki-code** [comment]
Ready.

File: xonsh/commands_cache.py [hunk 1]
```diff
@@ -75,41 +75,58 @@ def remove_dups(p):
                 ret.append(e)
         return ret
 
+    @staticmethod
+    def get_key(cmd):
+        return cmd.upper() if ON_WINDOWS else cmd
```
Comment (scopatz) [inline xonsh/commands_cache.py:80]: I don't think we need a new public API for this

Comment (anki-code) [reply xonsh/commands_cache.py]: reverted

**Comment by scopatz** [comment]
Seems like this has conflicts now @anki-code.

**Comment by anki-code** [comment]
Ready

File: xonsh/commands_cache.py [hunk 2]
```diff
@@ -83,23 +83,34 @@ def all_commands(self):
...
+            if cache_valid_aliases:
+                return self._cmds_cache
+            else:
+                for cmd, alias in alss.items():
+                    key = cmd.upper() if ON_WINDOWS else cmd
+                    if key in self._cmds_cache:
+                        self._cmds_cache[key] = (self._cmds_cache[key][0], alias)
+                    else:
+                        self._cmds_cache[key] = (cmd, True)
+                return self._cmds_cache
```
Comment (scopatz) [inline xonsh/commands_cache.py:112]: The return line can be moved outside of the if clause

File: xonsh/commands_cache.py [hunk 3]
```diff
@@ -83,23 +83,34 @@ def all_commands(self):
...
-        cache_valid = cache_valid and (max_mtime <= self._path_mtime)
+        cache_valid_paths = max_mtime <= self._path_mtime
         self._path_mtime = max_mtime
-        if cache_valid:
-            return self._cmds_cache
+
+        if cache_valid_path and cache_valid_paths:
+            if cache_valid_aliases:
+                return self._cmds_cache
+            else:
```
Comment (scopatz) [inline xonsh/commands_cache.py:105]: This could just be `if not cache_valid_aliases:`

**Comment by anki-code** [comment]
Ready

**Comment by scopatz** [comment]
Hey @anki-code - thanks for putting in this PR.  The code looks good to me, but I was having a hard time reproducing.  For me, 

```
scopatz@artemis:~$ xonsh --timings
 Debug level: Off
|------------------|-----------|-----------|
|Event name        | Time (s)  | Delta (s) |
|------------------|-----------|-----------|
|start             |   0.000   |   0.000   |
|pre_execer_init   |   0.011   |   0.011   |
|post_execer_init  |   0.046   |   0.035   |
|on_pre_rc         |   0.046   |   0.000   |
|on_post_rc        |   0.195   |   0.149   |
|on_post_init      |   0.204   |   0.009   |
|on_pre_cmdloop    |   0.204   |   0.000   |
|on_pre_prompt     |   0.204   |   0.000   |
|------------------|-----------|-----------|
```

So the `on_pre_prompt` diff is `0.000`. I have conda in my xonshrc.  Are you only seeing this issue on Windows WSL?

**Comment by anki-code** [comment]
Probably I'm not described the case step by step and with details. I try again.

1. I've found the issue in WSL and we added the warning about huge amount of files in pr #3911
2. Then I backed to pure Linux Ubuntu and found that CommandsCache (CC) works slow for me because it has two issues.
3. The first issue is that CC read the directory from PATH and the symlink to thisdirectory (double work). I've fixed it in pr #3915
4. The second issue is that CC invalidates cache and read files from all PATH directories every time when aliases have changed. I've caught this behavior on sourcing conda. The files from PATHs was scanned 3 times because conda adds the alias and invalidates the cache. This PR fix this. If there is no changes in PATH and PATH directories it will not scanned. As result I have x2 speed up the startup of the shell on Linux.

The case completely described in #3912.

**Comment by anki-code** [comment]
And finally I've found that this speed issue was not shown in `--timings` and I made the pr #3916 to fix it too.

**Comment by anki-code** [comment]
How prompt relates with CommandsCache (CC)? Prompt renders tokens i.e. git branch. To do this it calls the git command. The git command calls CC. And at the end of stack CC read the directories and files from PATH many times and slow because in linux there are around 3000 executable files in the PATH directories. I fixed this all.

**Comment by scopatz** [comment]
Thanks for the explanation! Just to be clear, you only saw the slow down on WSL though? I am in favor of the change, I just want to make sure

**Comment by anki-code** [comment]
The tables with timing in the first message of this thread are from pure Ubuntu:
* 0.738s - before this PR (reading of PATH directories 3 times)
* 0.389s - after this PR (reading of PATH directories 1 time)

**Comment by anki-code** [comment]
>  I was having a hard time reproducing. For me...

This is wrong time. If you want to test your xonsh timing:
1. Merge #3916 and post new table
2. Run [the code from 3915](https://github.com/xonsh/xonsh/issues/3915#issue-726199438) to get the amount of files in your PATH directories.
3. Install xonsh from master and run `COMMANDS_CACHE_SIZE_WARNING=1 xonsh --timings` and post here the count of warnings you see.

**Comment by anki-code** [comment]
Linux test

Before PR - 0.570:

<details>

```
(base) pc@pc:~$ COMMANDS_CACHE_SIZE_WARNING=1 xonsh --timings --rc ~/.xonshrc_conda 
Warning! Found 2,935 executable files in the PATH directories!
Warning! Found 2,935 executable files in the PATH directories!
Warning! Found 2,935 executable files in the PATH directories!
 Debug level: Off
|-------------------------------|-----------|-----------|
|Event name                     | Time (s)  | Delta (s) |
|-------------------------------|-----------|-----------|
|start                          |   0.000   |   0.000   |
|pre_execer_init                |   0.014   |   0.014   |
|post_execer_init               |   0.058   |   0.043   |
|on_pre_rc                      |   0.058   |   0.000   |
|on_post_rc                     |   0.233   |   0.175   |
|on_ptk_create                  |   0.313   |   0.080   |
|on_post_init                   |   0.313   |   0.000   |
|on_pre_cmdloop                 |   0.313   |   0.000   |
|on_pre_prompt                  |   0.313   |   0.000   |
|on_pre_prompt_tokens           |   0.345   |   0.032   |
|on_post_prompt_tokens          |   0.356   |   0.011   |
|on_pre_prompt_pygments_lexer   |   0.356   |   0.000   |
|on_post_prompt_pygments_lexer  |   0.570   |   0.214   |
|on_pre_prompter_prompt         |   0.570   |   0.000   |
|-------------------------------|-----------|-----------|

```

</details>

After PR - 0.392:

<details>

```
(base) pc@pc:~$ COMMANDS_CACHE_SIZE_WARNING=1 xonsh --timings --rc ~/.xonshrc_conda 
Warning! Found 2,935 executable files in the PATH directories!
 Debug level: Off
|-------------------------------|-----------|-----------|
|Event name                     | Time (s)  | Delta (s) |
|-------------------------------|-----------|-----------|
|start                          |   0.000   |   0.000   |
|pre_execer_init                |   0.015   |   0.015   |
|post_execer_init               |   0.059   |   0.044   |
|on_pre_rc                      |   0.059   |   0.000   |
|on_post_rc                     |   0.217   |   0.158   |
|on_ptk_create                  |   0.301   |   0.084   |
|on_post_init                   |   0.301   |   0.000   |
|on_pre_cmdloop                 |   0.301   |   0.000   |
|on_pre_prompt                  |   0.301   |   0.000   |
|on_pre_prompt_tokens           |   0.334   |   0.033   |
|on_post_prompt_tokens          |   0.349   |   0.015   |
|on_pre_prompt_pygments_lexer   |   0.349   |   0.000   |
|on_post_prompt_pygments_lexer  |   0.392   |   0.043   |
|on_pre_prompter_prompt         |   0.392   |   0.000   |
|-------------------------------|-----------|-----------|
```

</details>

The difference is not so big like in first message example because you already merge 3913 to the master.

**Comment by scopatz** [comment]
Ok, I'll go look at  #3916  first

**Comment by scopatz** [comment]
Thanks @anki-code - so on Ubuntu, I am seeing this:

**before PR**
<details>
```
scopatz@artemis:~$ xonsh --timings --shell-type=ptk
 Debug level: Off
|-------------------------|-----------|-----------|
|Event name               | Time (s)  | Delta (s) |
|-------------------------|-----------|-----------|
|start                    |   0.000   |   0.000   |
|pre_execer_init          |   0.011   |   0.011   |
|post_execer_init         |   0.048   |   0.036   |
|on_pre_rc                |   0.048   |   0.000   |
|on_post_rc               |   0.178   |   0.130   |
|on_ptk_create            |   0.254   |   0.076   |
|on_post_init             |   0.254   |   0.000   |
|on_pre_cmdloop           |   0.255   |   0.000   |
|on_pre_prompt_format     |   0.255   |   0.000   |
|on_pre_prompt_tokenize   |   0.277   |   0.023   |
|on_post_prompt_tokenize  |   0.288   |   0.011   |
|on_pre_prompt_style      |   0.288   |   0.000   |
|on_post_prompt_style     |   0.417   |   0.129   |
|on_pre_prompt            |   0.417   |   0.000   |
|-------------------------|-----------|-----------|
```
</details>

**After PR**
<details>
```
scopatz@artemis:~$ xonsh --timings --shell-type=ptk
 Debug level: Off
|-------------------------|-----------|-----------|
|Event name               | Time (s)  | Delta (s) |
|-------------------------|-----------|-----------|
|start                    |   0.000   |   0.000   |
|pre_execer_init          |   0.014   |   0.014   |
|post_execer_init         |   0.050   |   0.036   |
|on_pre_rc                |   0.050   |   0.000   |
|on_post_rc               |   0.180   |   0.130   |
|on_ptk_create            |   0.259   |   0.079   |
|on_post_init             |   0.259   |   0.000   |
|on_pre_cmdloop           |   0.259   |   0.000   |
|on_pre_prompt_format     |   0.259   |   0.000   |
|on_pre_prompt_tokenize   |   0.282   |   0.023   |
|on_post_prompt_tokenize  |   0.292   |   0.010   |
|on_pre_prompt_style      |   0.292   |   0.000   |
|on_post_prompt_style     |   0.439   |   0.147   |
|on_pre_prompt            |   0.439   |   0.000   |
|-------------------------|-----------|-----------|
base scopatz@artemis ~ $
```
</details>

So this is actually 5% slower than what is on master now.

**Comment by anki-code** [comment]
@scopatz this is with miniconda?

**Comment by scopatz** [comment]
Yeah, this is with miniconda in the xonshrc. Can you figure out a way to make it at least the same speed? Maybe skip the for loop if the size is small?

**Comment by anki-code** [comment]
Oh, I read "5% faster" but you wrote "5% slower". ROFL. I see that on my system the start 30% faster after PR. But you've got slower ))

**Comment by anki-code** [comment]
@scopatz take a look - https://asciinema.org/a/aiAebZ6LI0hV3nE43NwORhNfV

**Comment by anki-code** [comment]
>  Can you figure out a way to make it at least the same speed? Maybe skip the for loop if the size is small?

Take a look at the code a bit more deeeeper. It's the same and better than previous by speed.
In the old code the cache invalidation will raised every time. In the new code only if paths have changes.
The result you've got has difference around 0.02 are you kidding me? This is within the margin of error. =)

**Comment by scopatz** [comment]
I have rerun it many times and it is consistently slower on my system.  I have read this code many times now, and I am pretty sure that it is probably the for-loop that is slowing things down. I understand that this could be much faster in certain conditions. So what I am asking for here is that it be triggered in those conditions where we know the slow down happens. This seems like a reasonable first step to me, since it isn't always faster.

**Comment by scopatz** [comment]
I know this might seem very specific and particular, but I would like to get xonsh to boot in less than 0.1 seconds. So 0.02 is really, 20% of the startup time goal that I have.  

I know there are a lot of hurdles to jump over with this goal:

* Python's startup time
* Prompt-toolkit is slow to load
* Pygments is slow to load

But I hope that helps give some context on how I measure timing issues.

**Comment by anki-code** [comment]
>  I have read this code many times 

In the previous logic we see full cycle by PATHs and update all commands. Right?
In the new logic we see the for loop only by aliases that significantly smaller than PATHs. Right?
The question is. Why it's slower on 0.02s?

Could it be timing probes that added or another new features in the master?

**Comment by anki-code** [comment]
> in less than 0.1 seconds

It's a golden standard of web development and I completely understand you. We think similar.

**Comment by scopatz** [comment]
> The question is. Why it's slower on 0.02s?

Yeah, exactly! I don't really understand why it is slower ever :wink: 

Here is my xonshrc

<details>

```xsh
$EDITOR = 'xo'
$PROMPT = (
    '{env_name}{BOLD_GREEN}{user}@{hostname}{BOLD_BLUE} {cwd}{branch_color}'
    '{gitstatus: {}}{RESET} {BOLD_BLUE}{prompt_end}{RESET} '
)
$PROMPT_FIELDS['env_prefix'] = "{YELLOW}"
$PROMPT_FIELDS['env_postfix'] = " "

$QT_SCALE_FACTOR=2
#xontrib load mpl
#xontrib load xlsd
$XLSD_SORT_METHOD = "alphabetical"

import sys

$BASH_COMPLETION_USER_DIR = "/home/scopatz/miniforge/share/bash-completion"
aliases['git-authors'] = ['git', 'shortlog', '-s', '-n', '-e']

$MULTILINE_PROMPT = '`·.,¸,.·*¯`·.,¸,.·*¯'
$XONSH_SHOW_TRACEBACK = True
#$XONSH_STORE_STDOUT = True
$SHELL_TYPE = 'random'
#$SHELL_TYPE = 'readline'
#$SHELL_TYPE = 'prompt_toolkit'
#$UPDATE_COMPLETIONS_ON_KEYPRESS = True
#$UPDATE_PROMPT_ON_KEYPRESS = True
#aliases['curl'] = ['/usr/bin/curl']
#print('XONSH_DEBUG' in ${...})

$XONSH_COLOR_STYLE = 'monokai'

# sha256 conda forge
def _shatar(args, stdin=None):
    oldpwd = $PWD
    cd /tmp
    curl -L -O @("https://github.com/{0}/{1}/archive/{2}.tar.gz".format(*args))
    sha, _ = $(sha256sum @('{}.tar.gz'.format(args[2]))).split()
    print(sha)
    cd @(oldpwd)

aliases['shatar'] = _shatar

from xonsh.tools import unthreadable, uncapturable


def _mycat(args, stdin):
    for line in stdin:
        sys.stdout.write(line)
    sys.stdout.write('done with mycat!')

aliases['mycat'] = _mycat


def _feedstock_clone(args, stdin=None, stdout=None, stderr=None):
    """Clones a feedstock repo and sets up remotes"""
    from xonsh.lib.os import indir
    $package = args[0]
    git clone git@github.com:scopatz/$package-feedstock.git $package
    with indir($package):
        git remote add up git@github.com:conda-forge/$package-feedstock.git
        git fetch up
        git remote add bot git@github.com:regro-cf-autotick-bot/$package-feedstock.git
        git fetch bot
    del $package


aliases['feedstock-clone'] = _feedstock_clone

def _cf_mirror():
    """Mirrors conda-forge"""
    mounted = False
    if not p"~/mount/mirrors".exists():
        ![sudo mount /dev/sdb1 ~/mount/]
        mounted = True
    platforms = [
        "noarch",
        "linux-64",
        "osx-64",
        "osx-arm64",
        "win-64",
        "linux-ppc64le",
        "linux-aarch64",
    ]
    for platform in platforms:
        ![conda mirror --no-validate-target --upstream-channel conda-forge \
          --target-directory ~/mount/mirrors/conda-forge --temp-directory \
          ~/mount/tmp --platform @(platform) -vvv]
    # now run conda-suggest generate
    from xonsh.lib.os import indir
    with indir("/home/scopatz/conda-suggest-conda-forge/share/conda-suggest/"):
        ![conda suggest generate ~/mount/mirrors/conda-forge]
    ![du -h ~/mount/mirrors]
    if mounted:
        ![sudo umount ~/mount]


aliases["cf-mirror"] = _cf_mirror

# >>> conda initialize >>>
# !! Contents within this block are managed by 'conda init' !!
import sys as _sys
from types import ModuleType as _ModuleType
_mod = _ModuleType("xontrib.conda",
                   "Autogenerated from $(/home/scopatz/miniforge/bin/conda shell.xonsh hook)")
__xonsh__.execer.exec($("/home/scopatz/miniforge/bin/conda" "shell.xonsh" "hook"),
                      glbs=_mod.__dict__,
                      filename="$(/home/scopatz/miniforge/bin/conda shell.xonsh hook)")
_sys.modules["xontrib.conda"] = _mod
del _sys, _mod, _ModuleType
# <<< conda initialize <<<

aliases["gti"] = "git"
$PATH.insert(0, '/home/scopatz/.local/bin')
```

</details>

> Could it be timing probes that added or another new features in the master?

Good question! But I think it can't be, because my original numbers used master with the new timing probes. 


I'll try again with `--no-rc` and see if that changes anything.

**Comment by anki-code** [comment]
Don't forget to try `COMMANDS_CACHE_SIZE_WARNING=1 xonsh --timings`.
The amount of warnings will show how much times the full PATHs was read.

**Comment by scopatz** [comment]
here are my results:

Master:

<details>

```
scopatz@artemis:~$ xonsh --timings --shell-type=ptk --no-rc

                                Welcome to the xonsh shell (0.9.24.dev77)                                

                                  ~ All that is and all that shell be ~                                  

---------------------------------------------------------------------------------------------------------
xonfig tutorial    ->    Launch the tutorial in the browser
xonfig web         ->    Run the configuration tool in the browser and claim your shell 
(Note: Run the configuration tool or create a ~/.xonshrc file to suppress the welcome screen)

 Debug level: Off
|-------------------------|-----------|-----------|
|Event name               | Time (s)  | Delta (s) |
|-------------------------|-----------|-----------|
|start                    |   0.000   |   0.000   |
|pre_execer_init          |   0.016   |   0.016   |
|post_execer_init         |   0.050   |   0.035   |
|on_pre_rc                |   0.050   |   0.000   |
|on_post_rc               |   0.051   |   0.000   |
|on_ptk_create            |   0.119   |   0.068   |
|on_post_init             |   0.119   |   0.000   |
|on_pre_cmdloop           |   0.173   |   0.054   |
|on_pre_prompt_format     |   0.173   |   0.000   |
|on_pre_prompt_tokenize   |   0.173   |   0.000   |
|on_post_prompt_tokenize  |   0.270   |   0.097   |
|on_pre_prompt_style      |   0.270   |   0.000   |
|on_post_prompt_style     |   0.304   |   0.034   |
|on_pre_prompt            |   0.304   |   0.000   |
|-------------------------|-----------|-----------|
```

</details>

This PR: 

<details>

```
scopatz@artemis:~$ xonsh --timings --shell-type=ptk --no-rc

                                Welcome to the xonsh shell (0.9.24.dev86)                                

                                       ~ The shell, bourne again ~                                       

---------------------------------------------------------------------------------------------------------
xonfig tutorial    ->    Launch the tutorial in the browser
xonfig web         ->    Run the configuration tool in the browser and claim your shell 
(Note: Run the configuration tool or create a ~/.xonshrc file to suppress the welcome screen)

 Debug level: Off
|-------------------------|-----------|-----------|
|Event name               | Time (s)  | Delta (s) |
|-------------------------|-----------|-----------|
|start                    |   0.000   |   0.000   |
|pre_execer_init          |   0.011   |   0.011   |
|post_execer_init         |   0.047   |   0.036   |
|on_pre_rc                |   0.047   |   0.000   |
|on_post_rc               |   0.047   |   0.000   |
|on_ptk_create            |   0.112   |   0.065   |
|on_post_init             |   0.113   |   0.000   |
|on_pre_cmdloop           |   0.168   |   0.055   |
|on_pre_prompt_format     |   0.168   |   0.000   |
|on_pre_prompt_tokenize   |   0.168   |   0.000   |
|on_post_prompt_tokenize  |   0.263   |   0.096   |
|on_pre_prompt_style      |   0.263   |   0.000   |
|on_post_prompt_style     |   0.296   |   0.032   |
|on_pre_prompt            |   0.296   |   0.000   |
|-------------------------|-----------|-----------|
```

</details>

Huh, so that is faster with `--no-rc` by 2.6%....  It must be coming from my xonshrc then.

**Comment by scopatz** [comment]
To be more exact, something in my xonsh rc is triggering this to be slower on my system

**Comment by anki-code** [comment]
Don't forget to try `COMMANDS_CACHE_SIZE_WARNING=1 xonsh --timings` :)
The amount of warnings will show how much times the full PATHs was read.

**Comment by scopatz** [comment]
Ahh yeah, I am seeing:

```
Warning! Found 1,937 executable files in the PATH directories!
Warning! Found 2,132 executable files in the PATH directories!
```

**Comment by scopatz** [comment]
What is weird is that about 1 out of every 6 times, with the full xonshrc, I get a faster runtime of 0.406 s on this PR.

**Comment by anki-code** [comment]
May be it depends on SSD/HDD state?

**Comment by scopatz** [comment]
Yeah, I am on an SSD, FWIW

**Comment by anki-code** [comment]
I noticed something similar before this PR when xonsh every time reads PATHs 3 times at start. First two reads were fast and third was slow. After I install this PR I have one read and the speed is stable.

**Comment by anki-code** [comment]
Tests passed. Ready.

**Comment by scopatz** [comment]
Thanks @anki-code. Retesting it today, I didn't see any startup speed difference.

**Comment by anki-code** [comment]
@scopatz many thanks for this! After this PR + my PR #3913 I see 50% starting xonsh time reducing on my Ubuntu+conda. fyi @laloch my PRs not only about colors.

**Comment by laloch** [comment]
> I see 50% starting xonsh time reducing on my Ubuntu+conda.

That's great, thanks! I can't see any statistically significant difference here, but I'm not a conda user.

Regarding the colors, I was not talking about you specifically. I don't really keep track of who's committing what. I rather wanted to stress that "playing with colors" (which happens a lot recently) is completely different matter from changing operator semantics .

**Comment by anki-code** [comment]
Got it! Thank you for clarifications. 

As result we started process of reinventing as you wanted at the beginning :) and your opinion still welcome :)

**Pull request merged**